src/ngraph/placement.cpp

```diff
@@ -224,3 +224,186 @@ pair<vector<shared_ptr<Function>>, unordered_map<shared_ptr<op::Parameter>, shar
+        independent_nodes_by_placement.at(placement).pop_front();
+    }
+    else
+    {
+        for (auto& it : independent_nodes_by_placement)
+        {
+            if (it.second.size() > 0)
+            {
+                selected_node = it.second.front();
+                it.second.pop_front();
+                break;
+            }
+        }
+    }
+    return selected_node;
+}
+
+static vector<unordered_set<shared_ptr<Node>>>
+    group_function_nodes_to_clusters_size(const shared_ptr<Function>& f)
+{
+    // Topologically sort nodes by picking independent node with the same placement as the
+    // previously picked node greedily
+    map<size_t, deque<Node*>> independent_nodes_by_placement;
+    unordered_map<Node*, size_t> node_dependency_count;
+    unordered_map<ngraph::Node*, shared_ptr<ngraph::Node>> node_map;
+
+    for (shared_ptr<Node> node : f->get_ops())
+    {
+        size_t dependency_count = node->get_arguments().size();
+        node_map[node.get()] = node;
+        node_dependency_count[node.get()] = dependency_count;
+        if (dependency_count == 0)
+        {
+            independent_nodes_by_placement[node->get_placement_size()].push_back(node.get());
+        }
+    }
+
+    list<shared_ptr<Node>> sorted_nodes;
+    size_t previous_placement = 0; // Placement::DEFAULT
+    while (Node* independent_node = take_independent_node_with_placement_priority_size(
+               independent_nodes_by_placement, previous_placement))
+    {
+        previous_placement = independent_node->get_placement_size();
+        sorted_nodes.push_back(node_map.at(independent_node));
+
+        for (auto user : independent_node->get_users())
+        {
+            Node* user_node = user.get();
+            node_dependency_count.at(user_node) -= 1;
+            if (node_dependency_count.at(user_node) == 0)
+            {
+                independent_nodes_by_placement[user_node->get_placement_size()].push_back(
+                    user_node);
+            }
+        }
+    }
+
+    if (sorted_nodes.size() != f->get_ops().size())
+    {
+        throw ngraph_error("sorted_nodes.size()== " + to_string(sorted_nodes.size()) +
+                           " != f->get_ops().size()== " + to_string(f->get_ops().size()) +
+                           ". Internal error with topological sort.");
+    }
+
+    // Build clusters from the sorted_nodes
+    previous_placement = 0; // Placement::DEFAULT;
+    vector<unordered_set<shared_ptr<Node>>> clusters;
+    for (shared_ptr<Node> node : sorted_nodes)
+    {
+        size_t node_placement = node->get_placement_size();
+        if (node_placement != previous_placement)
+        {
+            unordered_set<shared_ptr<Node>> new_cluster;
+            clusters.push_back(new_cluster);
+        }
+        clusters.back().insert(node);
+        previous_placement = node_placement;
+    }
+
+    // Sanity check for node duplication and full node coverage
```

Comment:
nitpick: Can we move checks into standalone functions?? (`check_duplicates`, `check_coverage` This would make it much easier to read this function. It's already quite big

Reply:
if used more than once we can move it to a separate standalone function, if need I can make it a lamda ?